### ISSUE TITLE
FreeBSD target: service: name=postgresql state=started hangs the whole playbook

### ISSUE THREAD

**Issue by darklajid** [ansible/ansible#5923]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.4.4
##### Environment:

Arch Linux local, FreeBSD remote/target
##### Summary:

A very simple script hangs ansible-playground, making ansible unusable to deploy postgresql on FreeBSD without workarounds.
##### Steps To Reproduce:

Based on this playbook role:

```
tasks/main.yml:

---
- name: Install required packages
  pkgng: name=postgresql93-server,postgresql93-client state=present

- name: Install the postgresql rc.conf.d config file
  template: src=postgresql.rc.conf dest=/etc/rc.conf.d/postgresql owner=root group=wheel mode=0644

- name: Initialize the database cluster
  command: /usr/local/etc/rc.d/postgresql initdb creates=/usr/local/pgsql/data/pg_hba.conf 
  notify:
    - start postgresql

handlers/main.yml:

---
- name: start postgresql
  service: name=postgresql state=started
```

Run the playbook: ansible-playbook -K -i myinventory myentrypoint.yml
##### Expected Results:

ansible-playbook returns - either with success or an error message if something went wrong.
##### Actual Results:

ansible-playbook hangs forever.

```
TASK: [db | Install required packages] **************************************** 
changed: [ben.sh]

TASK: [db | Install the postgresql rc.conf.d config file] ********************* 
changed: [ben.sh]

TASK: [db | Initialize the database cluster] ********************************** 
changed: [ben.sh]

NOTIFIED: [ntp | restart ntpd] ************************************************ 
changed: [ben.sh]

NOTIFIED: [db | start postgresql] ********************************************* 
<=== Ansible hangs here ===>
```

The service is already running! If I abort ansible I get this stack:

```
^CTraceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 269, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/bin/ansible-playbook", line 209, in main
    pb.run()
  File "/usr/lib/python2.7/site-packages/ansible/playbook/__init__.py", line 263, in run
    if not self._run_play(play):
  File "/usr/lib/python2.7/site-packages/ansible/playbook/__init__.py", line 551, in _run_play
    self._run_task(play, handler, True)
  File "/usr/lib/python2.7/site-packages/ansible/playbook/__init__.py", line 364, in _run_task
    results = self._run_task_internal(task)
  File "/usr/lib/python2.7/site-packages/ansible/playbook/__init__.py", line 323, in _run_task_internal
    results = runner.run()
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 992, in run
    results = [ self._executor(h, None) for h in hosts ]
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 394, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 485, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 685, in _executor_internal_inner
    result = handler.run(conn, tmp, module_name, module_args, inject, complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/action_plugins/normal.py", line 54, in run
    return self.runner._execute_module(conn, tmp, module_name, module_args, inject=inject, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 360, in _execute_module
    res = self._low_level_exec_command(conn, cmd, tmp, sudoable=sudoable)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 771, in _low_level_exec_command
    rc, stdin, stdout, stderr = conn.exec_command(cmd, tmp, sudo_user, sudoable=sudoable, executable=executable)
  File "/usr/lib/python2.7/site-packages/ansible/runner/connection_plugins/ssh.py", line 221, in exec_command
    rfd, wfd, efd = select.select(rpipes, [], rpipes, 1)
KeyboardInterrupt
```

which seems to me like it's hanging in select? But I wouldn't know why..

Note that
- I have all the permissions I need, I run ansible-playbook -K and everything above is executed as root
- Starting services in general works fine (see the - unrelated to this problem - successful restart of ntpd in the log above)

The postgresql service itself is working fine. Both directly invoking the script at /usr/local/etc/rc.d/postgresql and via service postgresql works fine, and detaches immediately.

The only thing that I can think of that might cause issues(?): The service logs two lines before going to the background:

```
# service postgresql start
LOG:  ending log output to stderr
HINT:  Future log output will go to log destination "syslog".
```

Those messages are on stdout (not stderr) if that matters.

**Comment by darklajid** [comment]
Further update: Replacing the service: command in a handler with a simple command: service postgresql start in the task file results in the same problem. So there seems to be something fundamentally wrong in ansible's handling of service postgresql start here.

Running that locally on the FreeBSD machine (as root or using sudo) works flawlessly.
All tests were performed on a clean VM (rolling back to the starting conditions of 'just pkgng and python installed, otherwise bare' between each test).

**Comment by darklajid** [comment]
Final addendum:

These fail (playbook hangs forever, needs ctrl+c)

```
command: service postgresql start

shell: service postgresql start
```

These work (service starts and playbook continues):

```
raw: service postgresql start

shell: service postgresql start > /dev/null
```

At this point I think it is clear that this is a bug in the way ansible handles the command's output: A simple redirect to /dev/null 'solves' the problem and a raw command works fine as well.

**Comment by bcoca** [comment]
This sounds like the init script is not releasing the tty and keeping
stdout open, which is why ansible 'hangs'.

**Comment by mvk** [comment]
It might be due many reasons, e.g. the init script never starts (not on path) or waits for stdin.
you need to see what is actually being run by python on the other side.

**Comment by bcoca** [comment]
@MVK, init script is at:/usr/local/etc/rc.d/postgresql , which is part of
where the service module searches on FreeBSD, also if not found it would
return immediately with an error.

**Comment by darklajid** [comment]
mvk: The service module runs .. 'service' on FreeBSD, according to the ansible source.

Also, all my other samples explicitly list what is run:

```
service postgresql start
```

I would accept user error or mistakes in the init script, but I can run that locally, raw and with a redirection. What breaks the other uses (notably service:) and should that behavior be possible?

**Comment by darklajid** [comment]
@bcoca - keeping the tty wouldn't affect raw:?
Looking at the init script I don't see anything special..

**Comment by bcoca** [comment]
the raw module bypasses the normal module execution, it uses much simpler
path, which might account for the difference.

**Comment by darklajid** [comment]
So the question remains: Is this a bug in ansible? That script works locally just fine, I'm not sure (IF the init script is to blame) if a bug report a la 'works fine in general, but not if you happen to use ansible to call it' can succeed, to be honest.

**Comment by bcoca** [comment]
try executing the script remotely over ssh: ssh -t <machine> "service
postgresql start", I expect this to hang and the problem being that the
script isn't detaching itself from the tty correctly.

**Comment by darklajid** [comment]
On the road now, typing from my mobile.
Can give it a try in 6-8 hours.

Thanks!

**Comment by darklajid** [comment]
@bcoca - some tinc and host jumping tricks and fighting the android keyboard:

┌─[dar@TIS-Ben-T520]─[~]
└──╼ ssh -t 192.168.122.201 "sudo service postgresql start"
Enter passphrase for key '/home/dar/.ssh/id_rsa':
Password:
LOG:  ending log output to stderr
HINT:  Future log output will go to log destination "syslog".
Connection to 192.168.122.201 closed.
┌─[dar@TIS-Ben-T520]─[~]
└──╼

So, is that basically killing your assumption?

**Comment by darklajid** [comment]
FYI: Also happens with the current master (as of this writing):

```
ansible --version
ansible 1.5 (devel d3c1b8de78) last updated 2014/02/09 21:44:46 (GMT +200)
```

Same behavior.

**Comment by bcoca** [comment]
not sure this applies anymore, but worth a try (silent mode setting)

http://www.postgresql.org/message-id/A3BA2251DD85404FBBEF7478C29D8742932EC0@onmxm01.ciena.com

**Comment by swills** [comment]
Did a bunch of debugging on this. It's due the python subprocess popen communicate function waiting on the process to exit or close stdout (I think). Adding this:

postgresql_flags="-l /dev/null -w -s -m fast"

(adds -l /dev/null) to your templates/postgresql.rc.conf file should fix it.

**Comment by darklajid** [comment]
silent_mode was removed between 9.2 and 9.3 - so that doesn't apply anymore. But the release notes state (as Steve points out) that the same behavior can be configured by passing -l /some/file to pg_ctl - and he already points out how to do that for FreeBSD.

So I guess this is FreeBSD's or alternatively Postgresql's fault?

**Comment by swills** [comment]
I suppose in some sense it's Postresql's fault. But really, what ansible does is it's responsibility, and this could this could happen any time a daemon output's something at startup because the communicate call is reading the output then waiting for the process to end. See

http://docs.python.org/2/library/subprocess.html#subprocess.Popen.communicate

and

https://github.com/ansible/ansible/blob/5967b463c9e0d83ffbd958965a891ab4048205af/lib/ansible/module_utils/basic.py#L1014

Given systemd is encouraging app developers to write to stdout, I expect this will only become more of a problem, so it might be best for ansible to make some sort of change to the code, but I could be wrong.

**Comment by darklajid** [comment]
I tend to agree with you - at the end of the day ansible shouldn't hang. Especially since postgresql (in this case) works flawlessly, can be started locally or via ssh -t without any issue (see my comment further up this discussion). In other words: The postgresql init script doesn't pose a problem - unless you're using ansible. So I'd chalk this up as a corner case that exposes a weakness in the way ansible handles subprocesses and won't close this issue myself.

The python links don't help me all that much - I'm not sure what the subprocess is here. The python script on the target host? The 'sudo service postgresql start' call?

**Comment by darklajid** [comment]
Okay, the bug/issue is worse than I thought. The same happens with dovecot now.
In just two days I stumbled (it's not that I searched for these things, I'm just trying to build this simple mail environment) upon two services that cannot be handled with ansible, even block ansible forever.

I understand that a quick fix might be unlikely, but I'd be glad about any general 'Okay, yes that looks like a problem in ansible' sort of acknowledgement.

**Comment by jctanner** [comment]
@darklajid can we move this to a new feature request with a more accurate description? I think we have a root cause here, but my opinion is that it's a shortcoming of ansible, rather than a bug.

I would like to encourage anyone to contribute code to detect the subprocess call needs input.

**Comment by darklajid** [comment]
Hey. Thanks for getting back.

"Feature request vs bug": I .. don't know. But I don't think that I should argue the point. For me this is something that doesn't work and it should be fixed in ansible. If you prefer the 'well, it is just missing' view then it is a new feature I request. If you think that FreeBSD as supported target platform and service as a default/core module should 'just work' it might be a bug. But for me the distinction is useless, I guess. So - whatever you prefer.

Note that 'subprocess call needs input' is weird here, though. At this point I have to question if you're talking about the issue at hand? My issue (service state=started with postgresql or dovecot, probably more. I only used ansible to start 5? 6? services on FreeBSD so far, so the ratio is high) is not related to needing input, at least not for all I can tell.

Calling the command that the service module uses internally completes just fine, without any input by me. Issuing that command via ssh completes fine (and the ssh connection terminates as expected afterwards), without any input by me. Ansible hangs forever.

So while I've no trouble with the "is a bug just a missing feature in this case" part, the rest doesn't add up. Detecting whether a subprocess call needs input is .. irrelevant for the issue at hand, unless you know a lot more about the underlying ansible issue than this report contains so far.

**Comment by bcoca** [comment]
The problem normally stems from bad init script behavior (they really
should not need input), ansible tries to compensate for this but it is a
hard problem to solve. Hell, some are trying to throw away the whole system
(see systemd).

The bug/feature debate comes from 'should this be a init script bug or
should ansible compensate for them?'. IMHO both, which makes it a bug on
both sides, but I understand wanting to tag this as a 'feature' on the
ansible side.

**Comment by darklajid** [comment]
@bcoca I understand that the classification is more about politics and structure.

I still don't understand the issue though. You yourself suggested the ssh litmus test. If that passes, what's wrong with the init script? Shouldn't ansible behave exactly the same, shouldn't its 'send python over ssh to run a command' mechanism be transparent and result in the same thing?

Can you clarify what 'needs input' means here? I parsed that as 'user has to type something' so far, but that was obviously off. What's the issue here? You and @jctanner seem to know the cause for the reported behavior?

**Comment by bcoca** [comment]
The cause is not 100% clear, I'm mostly taking an educated guess that the
init script is not correctly detaching from the tty (this is the issue 99%
), normally easy to detect with my ssh test, but test mileage may vary.

I have used ansible to start/stop services on FreeBSD 8.x w/o any issues
(in my case Mysql not Postgresql) and I currently use it with Postgressql
on Ubuntu, but I have encountered issues with other daemons (tomcat,
selenium) and I was able to solve them by correctly daemonizing in the init
scripts. Since I have no experience with your current setup I cannot assure
this is the issue one way or the other, but I have a strong suspicion.

 I had hoped someone else with time and expertise would have looked at this
by now. If I find an opportunity to setup a similar environment I will
definitely test it, but my time is currently limited.

**Comment by jctanner** [comment]
Ok, let's just keep this open then. Once we know for sure what the real issue is, we can determine what to call it.

**Comment by darklajid** [comment]
If I can _help_ tracking this down: Let me know. Things I could do off the top of my head:
- help install FreeBSD 10-Release
- give access to a playground VPS with FreeBSD 10-Release

Unfortunately I'm neither decent at Python (I'm a programmer, can certainly read that but .. it's not home, it's unfamiliar), nor Ansible (just a user) nor FreeBSD (I'm exploring it for a server right now _with_ Ansible, haven't used it for a long time before that).

**Comment by swills** [comment]
The cause is 100% understood. Read my comment above. And this is a bug, not a feature request.

**Comment by bcoca** [comment]
@steve, sorry had not remembered your previous post, but that seems to
indicate the process outputs to stdin (many startup scripts do) but this
one never sends an EOF nor detaches.

**Comment by swills** [comment]
@bcoca it does output to stdout, which ansible receives on stdin, yes. And it does not send EOF more close it's stdout, but it does detach, at least in the case of postgresql93. Hence why the init script for the OS works fine. This is why I contend this is a bug in ansible.

**Comment by bcoca** [comment]
@steve,  most of my init script output to stdout and I don't see this
issue, that is why my suspicion was on something improper on the init
script.

Since you seem to have discounted that I'm perplexed on why this works/not
works inconsistently. I would expect to always fail or always work.

**Comment by darklajid** [comment]
To be safe: For me this fails always. Workarounds exist, but the core issue is easily reproducable in my environment.

**Comment by swills** [comment]
The issue is the postgresql93 subprocess rather than the init script itself. This is why adding "-l /dev/null" to the pg_ctl flags helps.

**Comment by Glandos** [comment]
Just to mention that this bug is hitting me hard with SLES11SP3 (the most up-to-date) and an init script for a custom service, based on the skeleton.
The call to communicate never ends, the playbook is stopped (or the shell/command module using ansible).
Redirecting stdout only to /dev/null doesn't work, I must add a redirection for stderr too (2>/dev/null).
It should be good to investigate.
My init script contains a call to su - $USER -c "/sbin/startproc $BIN" that may alter the behavior.

**Comment by darklajid** [comment]
Wait-a-sec. Can you expand 'SLES' here?
Are you saying you're having this issue on Linux?

Because that would change the scope of the bug quite a bit, I guess.

**Comment by Glandos** [comment]
Yes, I'm having this issue on SuSE Linux Enterprise (I think this is SLE-11 now, the last 'S' for 'server' disappeared some time ago).
I can give you the init script, but it will be truncated, since it's involving some private stuff.

**Comment by darklajid** [comment]
Unfortunately I'm not much help - I'm the original reporter, not part of the project. I do hope that the ansible crowd reconsiders the P4 label in light of your report.

A recent exchange elsewhere lead me to believe that the priority at the moment is P2. [1]
So unless you have an idea for a patch (I'm unable to dig deeper myself tbh) or the priority is changed, I guess we're stuck with that problem for a while.

1: https://news.ycombinator.com/item?id=8136619 - "there's a priority system for tagging tickets, where we hit P2 items first, and then some others. "

**Comment by mpdehaan** [comment]
Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

On September 26, 2014, due to enormous levels of contribution to the project Ansible decided to reorganize module repos, making it easier
for developers to work on the project and for us to more easily manage new contributions and tickets.

We split modules from the main project off into two repos, http://github.com/ansible/ansible-modules-core and http://github.com/ansible/ansible-modules-extras

If you would still like this ticket attended to, and believe this problem or idea is still present in the latest version of Ansible (1.7.2) or the development branch, we will need your help in having it reopened in one of the two new repos, and instructions are provided below.

We apologize that we are not able to make this transition happen seamlessly, though this is a one-time change and your help is greatly appreciated -- 
this will greatly improve velocity going forward.

Both sets of modules will ship with Ansible, though they'll receive slightly different ticket handling.

To locate where a module lives between 'core' and 'extras'
- Find the module at http://docs.ansible.com/list_of_all_modules.html
- Open the documentation page for that module
- If the bottom of the docs say "This is an extras module", submit your ticket to https://github.com/ansible/ansible-modules-extras
- Otherwise, submit your pull request to update the existing module to https://github.com/ansible/ansible-modules-core
- action_plugins (modules with server side components) still live in the main repo.  If your ticket affects both, open the ticket
   on the module repo just the same.

Additionally, should you need more help with this, you can ask questions on:
- the ansible-project mailing list: https://groups.google.com/forum/#!forum/ansible-project

Thank you very much!

**Comment by rmartin-chwy** [comment]
@darklajid , @Glandos , @bcoca, @swills : has this been fixed? Abandoned? I've encountered a very similar issue and opened it in ansible-modules-core, but have not gotten any assistance there as yet.

**Comment by wmoreno3** [comment]
Two issues solved for me, tks for comments I resumed.
server postgres[]: [2-1] FATAL: the database system is starting up
and
LOG:  ending log output to stderr
HINT:  Future log output will go to log destination "syslog".

I added `postgresql_flags="-l 2>/dev/null -s -m smart"` into /etc/rc.conf on FreeBSD 11.0 and PostgreSQL 9.6 and solved the issues.
`root@server:~ # service postgresql restart`
`waiting for server to shut down.... done`
`server stopped`
`server starting`
`root@server:~ #`
